### PR TITLE
KAFKA-16572: allow defining number of disks per broker in ClusterTest

### PR DESCRIPTION
**Pull request by FrankYang0529**
We introduced `disksPerBroker` in `TestKitNodes` from https://issues.apache.org/jira/browse/KAFKA-16559. We can support to config it in `ClusterTest`, so it's more convenient for integration tests.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java [hunk 1]
```diff
@@ -93,7 +93,7 @@ public List<Extension> getAdditionalExtensions() {
                 TestKitNodes nodes = new TestKitNodes.Builder().
                         setBootstrapMetadataVersion(clusterConfig.metadataVersion()).
                         setCombined(isCombined).
-                        setNumBrokerNodes(clusterConfig.numBrokers()).
+                        setNumBrokerNodes(clusterConfig.numBrokers(), clusterConfig.numDisksPerBroker()).
```
Comment (gaurav-narula) [inline core/src/test/java/kafka/test/junit/RaftClusterInvocationContext.java]:
```suggestion
                        setBrokerNodes(clusterConfig.numBrokers(), clusterConfig.numDisksPerBroker()).
```

File: core/src/test/java/kafka/test/ClusterTestExtensionsTest.java [hunk 2]
```diff
@@ -114,6 +118,25 @@ public void testClusterTests() {
...
+        result.allDescriptions().get().forEach((brokerId, logDirDescriptionMap) -> {
+            if (clusterInstance.clusterType().equals(ClusterInstance.ClusterType.ZK)) {
+                Assertions.assertEquals(1, logDirDescriptionMap.size());
+            } else if (clusterInstance.clusterType().equals(ClusterInstance.ClusterType.RAFT)) {
+                Assertions.assertEquals(2, logDirDescriptionMap.size());
+            } else {
+                Assertions.fail("Unknown cluster type " + clusterInstance.clusterType());
+            }
+        });
+    }
```
Comment (gaurav-narula) [inline core/src/test/java/kafka/test/ClusterTestExtensionsTest.java:125]: Nit: new line after the function

File: core/src/test/java/kafka/testkit/TestKitNodes.java [hunk 3]
```diff
@@ -82,6 +82,10 @@ public Builder setNumBrokerNodes(int numBrokerNodes) {
             return setBrokerNodes(numBrokerNodes, 1);
         }
 
+        public Builder setNumBrokerNodes(int numBrokerNodes, int disksPerBroker) {
```
Comment (chia7712) [inline core/src/test/java/kafka/testkit/TestKitNodes.java]: why we need this method? it is equal to `setBrokerNodes`

File: core/src/test/java/kafka/test/ClusterConfig.java [hunk 4]
```diff
@@ -155,29 +161,31 @@ public ClusterConfig copyOf() {
...
+        private int disksPerBroker;
         private String name;
         private boolean autoStart;
         private SecurityProtocol securityProtocol;
         private String listenerName;
         private File trustStoreFile;
         private MetadataVersion metadataVersion;
 
-        Builder(Type type, int brokers, int controllers, boolean autoStart, SecurityProtocol securityProtocol, MetadataVersion metadataVersion) {
+        Builder(Type type, int brokers, int controllers, int disksPerBroker, boolean autoStart, SecurityProtocol securityProtocol, MetadataVersion metadataVersion) {
```
Comment (chia7712) [inline core/src/test/java/kafka/test/ClusterConfig.java]: Could we call setters instead of having a complicated constructor? We should not waste the build pattern ... this is unrelated to this PR, and it already have a jira (https://issues.apache.org/jira/browse/KAFKA-16560). However, we can do a bit refactor here.

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 thanks for enhancing the infra. two minor comments left.

**Comment by FrankYang0529** [comment]
Hi @gaurav-narula and @chia7712, I have addressed all comments. Thanks for your review.

File: core/src/test/java/kafka/test/junit/ClusterTestExtensions.java [hunk 5]
```diff
@@ -179,8 +186,8 @@ private void processClusterTest(ExtensionContext context, ClusterTest annot, Clu
                 throw new IllegalStateException();
         }
 
-        ClusterConfig.Builder builder = ClusterConfig.clusterBuilder(type, brokers, controllers, autoStart,
-            annot.securityProtocol(), annot.metadataVersion());
+        ClusterConfig.Builder builder = ClusterConfig.clusterBuilder(type, brokers, controllers, disksPerBroker,
```
Comment (chia7712) [inline core/src/test/java/kafka/test/junit/ClusterTestExtensions.java]: It seems to me `clusterBuilder` should be removed. Also, the temporary local variables (`type`, `brokers`, etc ) can be removed. We have a builder already and hence we should leverage it to collect all variables. for example:
```java
        ClusterConfig.Builder builder = ClusterConfig.builder();
        if (annot.clusterType() == Type.DEFAULT) {
            builder.type(defaults.clusterType());
        } else {
            builder.type(annot.clusterType());
        }
```

Comment (FrankYang0529) [reply core/src/test/java/kafka/test/junit/ClusterTestExtensions.java]: Do we want to use another jira to trace this? Or I remove it in this PR? Thanks.

Comment (chia7712) [reply core/src/test/java/kafka/test/junit/ClusterTestExtensions.java]: We can do that in this PR as most changes involved by refactor are in this method. Let's do it to decorate our test infrastructure:)

Comment (FrankYang0529) [reply core/src/test/java/kafka/test/junit/ClusterTestExtensions.java]: Thanks for the suggestion. I updated it.

**Comment by chia7712** [comment]
@FrankYang0529 This PR will enhance our new test infra. However, @brandboat is working at refactoring the infra at the same time. It seems to me this PR should be blocked by #15761 in order to avoid big conflicts. Also, this PR can do the enhancement based on the new graceful code base.

**Comment by chia7712** [comment]
@FrankYang0529 please fix the conflicts

**Comment by FrankYang0529** [comment]
> @FrankYang0529 please fix the conflicts

Hi @chia7712, I resolve conflicts. Thank you.

File: core/src/test/java/kafka/test/ClusterConfig.java [hunk 6]
```diff
@@ -160,6 +166,7 @@ public boolean equals(Object object) {
         return Objects.equals(type, clusterConfig.type)
                 && Objects.equals(brokers, clusterConfig.brokers)
                 && Objects.equals(controllers, clusterConfig.controllers)
+                && Objects.equals(disksPerBroker, clusterConfig.disksPerBroker)
```
Comment (chia7712) [inline core/src/test/java/kafka/test/ClusterConfig.java]: not sure whether we need to keep `hashCode` and `equals`. We don't use it actually and it is error-prone when adding new fields. For example, this PR does not update `hashCode` for the new field. @brandboat WDTY?

File: core/src/test/java/kafka/test/junit/ClusterTestExtensions.java [hunk 7]
```diff
@@ -162,32 +162,39 @@ private void processClusterTest(ExtensionContext context, ClusterTest annot, Clu
             controllers = annot.controllers();
         }
 
-        if (brokers <= 0 || controllers <= 0) {
-            throw new IllegalArgumentException("Number of brokers/controllers must be greater than zero.");
+        final int disksPerBroker;
+        if (annot.disksPerBroker() == 0) {
+            disksPerBroker = defaults.disksPerBroker();
```
Comment (chia7712) [inline core/src/test/java/kafka/test/junit/ClusterTestExtensions.java]: Could we leverage the builder instead of creating a bunch of temporary local variables?

File: core/src/test/java/kafka/test/ClusterConfig.java [hunk 8]
```diff
@@ -55,14 +56,15 @@ public class ClusterConfig {
...
-    private ClusterConfig(Type type, int brokers, int controllers, String name, boolean autoStart,
+    private ClusterConfig(Type type, int brokers, int controllers, int disksPerBroker, String name, boolean autoStart,
                   SecurityProtocol securityProtocol, String listenerName, File trustStoreFile,
                   MetadataVersion metadataVersion, Map<String, String> serverProperties, Map<String, String> producerProperties,
                   Map<String, String> consumerProperties, Map<String, String> adminClientProperties, Map<String, String> saslServerProperties,
                   Map<String, String> saslClientProperties, Map<Integer, Map<String, String>> perBrokerOverrideProperties) {
         this.type = Objects.requireNonNull(type);
         this.brokers = brokers;
         this.controllers = controllers;
+        this.disksPerBroker = disksPerBroker;
```
Comment (chia7712) [inline core/src/test/java/kafka/test/ClusterConfig.java:72]: Could you add check to make sure the value is larger than zero?

File: core/src/test/java/kafka/test/ClusterTestExtensionsTest.java [hunk 9]
```diff
@@ -121,6 +124,26 @@ public void testClusterTests() {
         }
     }
 
+    @ClusterTests({
+        @ClusterTest(clusterType = Type.ZK, disksPerBroker = 1),
```
Comment (chia7712) [inline core/src/test/java/kafka/test/ClusterTestExtensionsTest.java]: We can't use `disksPerBroker = 1` since it is equal to the default value.

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 nice improvement! a couple of comments left.

File: core/src/test/java/kafka/test/ClusterConfig.java (hunk 6 above)
Comment (brandboat) [reply core/src/test/java/kafka/test/ClusterConfig.java]: feel free to remove them as currently they are unused 😃

Comment (FrankYang0529) [reply core/src/test/java/kafka/test/ClusterConfig.java]: Remove `hashCode` and `equals` functions. Thanks.

File: core/src/test/java/kafka/test/junit/ClusterTestExtensions.java (hunk 7 above)
Comment (FrankYang0529) [reply core/src/test/java/kafka/test/junit/ClusterTestExtensions.java]: Yeah, I remove `brokers`, `controllers`, and `disksPerBroker`, but keeping `type`. We can't get attribute from `ClusterConfig#Builder` and we need `type` to run `invocationContexts`.

File: core/src/test/java/kafka/test/ClusterConfig.java (hunk 8 above)
Comment (FrankYang0529) [reply core/src/test/java/kafka/test/ClusterConfig.java:72]: I add check in `ClusterConfig#Builder#build` function.

File: core/src/test/java/kafka/test/ClusterTestExtensionsTest.java (hunk 9 above)
Comment (FrankYang0529) [reply core/src/test/java/kafka/test/ClusterTestExtensionsTest.java]: Thanks for finding this. Remove it.

**Review by showuon — APPROVED** [review]
LGTM!

File: core/src/test/java/kafka/test/junit/ClusterTestExtensions.java (hunk 7 above)
Comment (chia7712) [reply core/src/test/java/kafka/test/junit/ClusterTestExtensions.java]: We can leverage ternary conditional operator to make it more simple. For example:
```java
        Type type = annot.clusterType() == Type.DEFAULT ? defaults.clusterType() : annot.clusterType();
        Map<String, String> serverProperties = new HashMap<>();
        for (ClusterConfigProperty property : defaults.serverProperties()) {
            serverProperties.put(property.key(), property.value());
        }
        for (ClusterConfigProperty property : annot.serverProperties()) {
            serverProperties.put(property.key(), property.value());
        }

        ClusterConfig config = ClusterConfig.builder()
            .setType(type)
            .setBrokers(annot.brokers() == 0 ? defaults.brokers() : annot.brokers())
            .setControllers(annot.controllers() == 0 ? defaults.controllers() : annot.controllers())
            .setDisksPerBroker(annot.disksPerBroker() == 0 ? defaults.disksPerBroker() : annot.disksPerBroker())
            .setAutoStart(annot.autoStart() == AutoStart.DEFAULT ? defaults.autoStart() : annot.autoStart() == AutoStart.YES)
            .setName(annot.name().isEmpty() ? null : annot.name())
            .setName(annot.listener().isEmpty() ? null : annot.listener())
            .setServerProperties(serverProperties)
            .setSecurityProtocol(annot.securityProtocol())
            .setMetadataVersion(annot.metadataVersion())
            .build();

        type.invocationContexts(context.getRequiredTestMethod().getName(), config, testInvocations);
```

Comment (FrankYang0529) [reply core/src/test/java/kafka/test/junit/ClusterTestExtensions.java]: Thanks for the suggestion. It's more clear. Updated it.

File: core/src/test/java/kafka/test/ClusterConfig.java [hunk 10]
```diff
@@ -319,7 +302,10 @@ public Builder setPerBrokerProperties(Map<Integer, Map<String, String>> perBroke
         }
 
         public ClusterConfig build() {
-            return new ClusterConfig(type, brokers, controllers, name, autoStart, securityProtocol, listenerName,
+            if (brokers <= 0 || controllers <= 0 || disksPerBroker <= 0) {
```
Comment (chia7712) [inline core/src/test/java/kafka/test/ClusterConfig.java]: We should enable to set `brokers` to 0 as it is valid to have a cluster with only quorum nodes.

Comment (chia7712) [reply core/src/test/java/kafka/test/ClusterConfig.java]: btw, those check should be moved to constructor of `ClusterConfig`

File: core/src/test/java/kafka/test/junit/ClusterTestExtensions.java [hunk 11]
```diff
@@ -202,8 +151,19 @@ private void processClusterTest(ExtensionContext context, ClusterTest annot, Clu
...
         }
-        configBuilder.setServerProperties(serverProperties);
-        type.invocationContexts(context.getRequiredTestMethod().getName(), configBuilder.build(), testInvocations);
+        ClusterConfig config = ClusterConfig.builder()
+                .setType(type)
+                .setBrokers(annot.brokers() == 0 ? defaults.brokers() : annot.brokers())
+                .setControllers(annot.controllers() == 0 ? defaults.controllers() : annot.controllers())
+                .setDisksPerBroker(annot.disksPerBroker() == 0 ? defaults.disksPerBroker() : annot.disksPerBroker())
+                .setAutoStart(annot.autoStart() == AutoStart.DEFAULT ? defaults.autoStart() : annot.autoStart() == AutoStart.YES)
+                .setName(annot.name().isEmpty() ? null : annot.name())
```
Comment (chia7712) [inline core/src/test/java/kafka/test/junit/ClusterTestExtensions.java]: maybe `annot.name().trim().isEmpty()` is more suitable.

File: core/src/test/java/kafka/test/junit/ClusterTestExtensions.java [hunk 12]
```diff
@@ -202,8 +151,19 @@ private void processClusterTest(ExtensionContext context, ClusterTest annot, Clu
...
+                .setListenerName(annot.listener().isEmpty() ? null : annot.listener())
```
Comment (chia7712) [inline core/src/test/java/kafka/test/junit/ClusterTestExtensions.java]: ditto

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 thanks for updating code. some comments left. PTAL

**Comment by FrankYang0529** [comment]
Hi @chia7712, thanks for the review. I address all comments and add some test cases for ClusterConfig.

File: core/src/test/java/kafka/test/ClusterConfig.java [hunk 13]
```diff
@@ -55,14 +56,21 @@ public class ClusterConfig {
...
-    private ClusterConfig(Type type, int brokers, int controllers, String name, boolean autoStart,
+    private ClusterConfig(Type type, int brokers, int controllers, int disksPerBroker, String name, boolean autoStart,
                   SecurityProtocol securityProtocol, String listenerName, File trustStoreFile,
                   MetadataVersion metadataVersion, Map<String, String> serverProperties, Map<String, String> producerProperties,
                   Map<String, String> consumerProperties, Map<String, String> adminClientProperties, Map<String, String> saslServerProperties,
                   Map<String, String> saslClientProperties, Map<Integer, Map<String, String>> perBrokerOverrideProperties) {
+        if (brokers < 0) {
+            throw new IllegalArgumentException("Number of brokers must be greater or equal to zero.");
+        }
+        if (controllers <= 0 || disksPerBroker <= 0) {
```
Comment (chia7712) [inline core/src/test/java/kafka/test/ClusterConfig.java]: `controllers <= 0` is acceptable if it is zk mode. Also, `TestKitNodes` guards against that already

https://github.com/apache/kafka/blob/89d8045a15b622805f65c3c6fbfde82606921f65/core/src/test/java/kafka/testkit/TestKitNodes.java#L90

Hence, `ClusterConfig` does not require such check, and we can add comment to explain why we don't need to check `controllers` here.

File: core/src/test/java/kafka/test/ClusterTestExtensionsTest.java [hunk 14]
```diff
@@ -104,6 +107,26 @@ public void testClusterTests() {
         }
     }
 
+    @ClusterTests({
+        @ClusterTest(clusterType = Type.ZK),
```
Comment (chia7712) [inline core/src/test/java/kafka/test/ClusterTestExtensionsTest.java:111]: We should test all types for both cases: "default" + "custom"

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 thanks for updated PR. two comments left.

**Comment by FrankYang0529** [comment]
> @FrankYang0529 thanks for updated PR. two comments left.

Hi @chia7712, thanks for the review. I addressed last comments.

File: core/src/test/java/kafka/test/ClusterConfig.java (hunk 13 above)
Comment (chia7712) [reply core/src/test/java/kafka/test/ClusterConfig.java]: My point was "controllers can be zero in zk case", so `ClusterConfig` should not throw exception. Hence, we should have following checks.
```java
        // do fail fast. the following values are invalid for both zk and kraft modes.
        if (brokers < 0) throw new IllegalArgumentException("Number of brokers must be greater or equal to zero.");
        if (controllers < 0) throw new IllegalArgumentException("Number of controller must be greater or equal to zero.");
        if (disksPerBroker <= 0) throw new IllegalArgumentException("Number of disks must be greater than zero.");
```

Comment (FrankYang0529) [reply core/src/test/java/kafka/test/ClusterConfig.java]: Sorry, I misunderstood your message. Updated it. Thank you.

**Review by chia7712 — APPROVED** [review]
LGTM
